File: elixir/src/json.ex
# Minimal JSON support for sekreto.
#
# sekreto adds no third-party dependencies, so it carries just enough JSON
# to read a vault's answer and write the CLI's own line of output. It is
# deliberately not a general-purpose library.
#
# OTP 27 ships `:json` and Elixir 1.18 a `JSON` module. Neither is used,
# and neither is reached for conditionally: the OTP 25 that Ubuntu 24.04
# packages has no built-in JSON at all, so a port that used one where it
# found one would behave differently on two releases of the same runtime.
#
# Tagged tuples rather than plain Elixir values: a vault answering `null`,
# `false`, `0` and "no such key" means four different things, and a closed
# value model keeps them apart rather than by convention. `parse` answers
# `{:ok, value}` or `:error`, where `:error` means "this text is not JSON"
# and `{:ok, :null}` means "this text is the JSON literal null" - a
# distinction the callers of fetchjson need, since only the first is a
# malformed response.
#
# Objects are an ORDERED list of pairs, not a map: a payload's field order
# is signed, and Elixir maps have no order at all above 32 keys.
#
# A port of typescript/src/Json.ts, which is canonical.

defmodule Sekreto.Json do
  @moduledoc "The JSON value model, parser and writer sekreto carries."

  # A response body arrives before any trust check has run, so `[[[[...`
  # must not be able to exhaust the stack.
  @maxdepth 128

  @typedoc """
  A JSON value. `:none` is not one: it is what the accessors below answer
  when there is nothing there at all, which JSON cannot say.
  """
  @type t ::
          :null
          | {:bool, boolean}
          | {:num, float}
          | {:str, binary}
          | {:arr, [t]}
          | {:obj, [{binary, t}]}

  # ------------------------------------------------------- constructors

  def str(value), do: {:str, value}

  def num(value), do: {:num, value / 1}

  def bool(value), do: {:bool, value}

  def arr(entries), do: {:arr, entries}

  @doc "An object, in the order given: a payload's field order is signed."
  def obj(pairs), do: {:obj, pairs}

  # ---------------------------------------------------------- accessors
  #
  # Each answers `:none` for the wrong type as well as for a missing value,
  # because every caller wants the same thing from both: `__type` must be a
  # string, a 1Password vault list must be an array, a Doppler config must
  # be an object.

  def asstr({:str, value}), do: value
  def asstr(_other), do: :none

  def asnum({:num, value}), do: value
  def asnum(_other), do: :none

  def asarr({:arr, value}), do: value
  def asarr(_other), do: :none

  def asobj({:obj, value}), do: value
  def asobj(_other), do: :none

  @doc "Walk nested objects; `:none` the moment a step is not there."
  def dig(value, keys) when is_list(keys) do
    Enum.reduce(keys, value, fn key, at ->
      case at do
        {:obj, pairs} ->
          case List.keyfind(pairs, key, 0) do
            {^key, found} -> found
            nil -> :none
          end

        _other ->
          :none
      end
    end)
  end

  @doc """
  This value as the text a caller would print, or `:none` when there is no
  value at all.

  A JSON null is "no value": every provider here treats it as a miss rather
  than as the string "null".
  """
  def text(:null), do: :none
  def text(:none), do: :none
  def text({:str, value}), do: value
  def text({:num, value}), do: numstr(value)
  def text({:bool, value}), do: if(value, do: "true", else: "false")
  def text(other), do: stringify(other)

  # ------------------------------------------------------------- writer

  @doc "Render a value as compact JSON."
  def stringify(:null), do: "null"
  def stringify(:none), do: "null"
  def stringify({:bool, value}), do: if(value, do: "true", else: "false")
  def stringify({:num, value}), do: numstr(value)
  def stringify({:str, value}), do: quotestr(value)
  def stringify({:arr, entries}), do: "[" <> Enum.map_join(entries, ",", &stringify/1) <> "]"

  def stringify({:obj, pairs}) do
    "{" <> Enum.map_join(pairs, ",", fn {key, value} -> quotestr(key) <> ":" <> stringify(value) end) <> "}"
  end

  @doc """
  Render a string as a JSON string literal, quotes included.

  Public because the CLI assembles its one line of output field by field:
  printing a map there is what has bitten port after port, since the
  language's own key order is not the one every other port prints.
  """
  def quotestr(text) when is_binary(text) do
    "\"" <> escape(text, []) <> "\""
  end

  defp escape(<<>>, acc), do: acc |> Enum.reverse() |> IO.iodata_to_binary()

  defp escape(<<ch::utf8, rest::binary>>, acc) do
    piece =
      case ch do
        ?" -> "\\\""
        ?\\ -> "\\\\"
        ?\n -> "\\n"
        ?\r -> "\\r"
        ?\t -> "\\t"
        # `/` is accepted on input but never escaped on output, and
        # non-ASCII is not escaped either.
        other when other < 0x20 -> "\\u" <> String.pad_leading(Integer.to_string(other, 16), 4, "0")
        other -> <<other::utf8>>
      end

    escape(rest, [piece | acc])
  end

  @doc """
  Render a number the way every other port does: a whole number has no
  fractional tail, so a JSON `1` read back and printed stays `1`.
  """
  def numstr(value) do
    asfloat = value / 1

    if asfloat == Float.round(asfloat) and 9_007_199_254_740_992.0 > abs(asfloat) do
      Integer.to_string(trunc(asfloat))
    else
      Float.to_string(asfloat)
    end
  end

  # ------------------------------------------------------------- parser

  @doc """
  Parse JSON text.

  `:error` for anything unreadable - which the caller must tell apart from
  a literal `null` body, since only the first means the store could not
  answer coherently.
  """
  def parse(text) when is_binary(text) and byte_size(text) > 0 do
    # No parse error escapes: a malformed body is a fact about the body,
    # not an exception the caller has to be ready for.
    try do
      {value, rest} = pvalue(skip(text), 0)

      case skip(rest) do
        <<>> -> {:ok, value}
        _more -> :error
      end
    rescue
      _err -> :error
    catch
      :badjson -> :error
    end
  end

  def parse(_other), do: :error

  defp skip(<<ch, rest::binary>>) when ch in [?\s, ?\t, ?\n, ?\r], do: skip(rest)
  defp skip(text), do: text

  defp deeper(depth) do
    if @maxdepth <= depth, do: throw(:badjson)
    depth + 1
  end

  defp pvalue(<<?{, rest::binary>>, depth), do: pobj(skip(rest), deeper(depth), [])
  defp pvalue(<<?[, rest::binary>>, depth), do: parr(skip(rest), deeper(depth), [])
  defp pvalue(<<?", rest::binary>>, _depth), do: pstr(rest, [])
  # Matched whole, never by first letter.
  defp pvalue(<<"true", rest::binary>>, _depth), do: {{:bool, true}, rest}
  defp pvalue(<<"false", rest::binary>>, _depth), do: {{:bool, false}, rest}
  defp pvalue(<<"null", rest::binary>>, _depth), do: {:null, rest}
  defp pvalue(text, _depth), do: pnum(text)

  defp pobj(<<?}, rest::binary>>, _depth, []), do: {{:obj, []}, rest}

  defp pobj(text, depth, acc) do
    {{:str, key}, rest} =
      case skip(text) do
        <<?", body::binary>> -> pstr(body, [])
        _other -> throw(:badjson)
      end

    rest =
      case skip(rest) do
        <<?:, more::binary>> -> more
        _other -> throw(:badjson)
      end

    {value, rest} = pvalue(skip(rest), depth)
    acc = [{key, value} | acc]

    case skip(rest) do
      <<?,, more::binary>> -> pobj(skip(more), depth, acc)
      <<?}, more::binary>> -> {{:obj, Enum.reverse(acc)}, more}
      _other -> throw(:badjson)
    end
  end

  defp parr(<<?], rest::binary>>, _depth, []), do: {{:arr, []}, rest}

  defp parr(text, depth, acc) do
    {value, rest} = pvalue(skip(text), depth)
    acc = [value | acc]

    case skip(rest) do
      <<?,, more::binary>> -> parr(skip(more), depth, acc)
      <<?], more::binary>> -> {{:arr, Enum.reverse(acc)}, more}
      _other -> throw(:badjson)
    end
  end

  defp pstr(<<?", rest::binary>>, acc) do
    {{:str, acc |> Enum.reverse() |> IO.iodata_to_binary()}, rest}
  end

  defp pstr(<<?\\, esc, rest::binary>>, acc) do
    case esc do
      ?" -> pstr(rest, ["\"" | acc])
      ?\\ -> pstr(rest, ["\\" | acc])
      ?/ -> pstr(rest, ["/" | acc])
      ?b -> pstr(rest, ["\b" | acc])
      ?f -> pstr(rest, ["\f" | acc])
      ?n -> pstr(rest, ["\n" | acc])
      ?r -> pstr(rest, ["\r" | acc])
      ?t -> pstr(rest, ["\t" | acc])
      ?u -> punicode(rest, acc)
      _other -> throw(:badjson)
    end
  end

  defp pstr(<<>>, _acc), do: throw(:badjson)

  defp pstr(<<ch::utf8, rest::binary>>, acc), do: pstr(rest, [<<ch::utf8>> | acc])

  # An unpaired byte is still data; a vault answering latin-1 is malformed,
  # not a reason to crash.
  defp pstr(<<ch, rest::binary>>, acc), do: pstr(rest, [<<ch>> | acc])

  # `\uXXXX` carries a UTF-16 code unit. A surrogate PAIR is recombined
  # here, unlike the JVM ports, which append both units to a UTF-16 string
  # and get the same character for free - an Elixir binary is UTF-8 and has
  # no way to hold half a character. A lone surrogate, which no correct
  # encoder emits, becomes the replacement character rather than raising.
  defp punicode(text, acc) do
    {high, rest} = hex4(text)

    cond do
      0xD800 <= high and 0xDBFF >= high ->
        case rest do
          <<?\\, ?u, more::binary>> ->
            {low, after_low} = hex4(more)

            if 0xDC00 <= low and 0xDFFF >= low do
              code = 0x10000 + (high - 0xD800) * 0x400 + (low - 0xDC00)
              pstr(after_low, [<<code::utf8>> | acc])
            else
              pstr(after_low, [<<0xFFFD::utf8>>, <<0xFFFD::utf8>> | acc])
            end

          _other ->
            pstr(rest, [<<0xFFFD::utf8>> | acc])
        end

      0xDC00 <= high and 0xDFFF >= high ->
        pstr(rest, [<<0xFFFD::utf8>> | acc])

      true ->
        pstr(rest, [<<high::utf8>> | acc])
    end
  end

  defp hex4(<<digits::binary-size(4), rest::binary>>) do
    case Integer.parse(digits, 16) do
      {value, ""} -> {value, rest}
      _other -> throw(:badjson)
    end
  end

  defp hex4(_short), do: throw(:badjson)

  defp pnum(text) do
    taken = numlen(text, 0)

    if 0 == taken, do: throw(:badjson)

    <<span::binary-size(taken), rest::binary>> = text

    # Float.parse refuses a non-finite result outright, which is what JSON
    # wants: `1e999` is not a number this library can carry, and letting it
    # through as an infinity blows up a token-expiry computation later.
    case Float.parse(span) do
      {value, ""} -> {{:num, value}, rest}
      _other -> throw(:badjson)
    end
  end

  defp numlen(text, at) do
    case text do
      <<_::binary-size(at), ch, _::binary>> when ch in ?0..?9 or ch in [?-, ?+, ?., ?e, ?E] ->
        numlen(text, at + 1)

      _other ->
        at
    end
  end
end

File: swift/src/Json.swift
// Minimal JSON support for sekreto.
//
// sekreto adds no third-party dependencies, so it carries just enough JSON
// to read a vault's answer and write the CLI's own line of output. It is
// deliberately not a general-purpose library.
//
// Hand-rolled rather than JSONSerialization, for two reasons that both
// matter here. JSONSerialization hands back an unordered Dictionary, and a
// signed payload's field order is part of what was signed; and it bridges
// numbers through NSNumber, where `true` and `1` are not reliably
// distinguishable. An `enum` keeps a vault answering `null`, `false`, `0`
// and "no such key" as four different things at compile time rather than
// by convention.
//
// `parse` returns `Json?`, where `nil` means "this text is not JSON" and
// `.null` means "this text is the JSON literal null" - a distinction the
// plugins' HTTP round-trip needs, since only the first is a malformed
// response. It stays in the core because the core writes JSON too: a
// Sekreto describes itself as data, without its secrets.
//
// A port of typescript/src/Json.ts, which is canonical.

import Foundation

/// An insertion-ordered map.
///
/// Swift's `Dictionary` is unordered, and three things here are compared or
/// signed as whole maps: a memory provider's `values`, the SigV4 output
/// headers, and a `.env` file's contents. A pair list keeps the order the
/// author wrote, at a lookup cost that is irrelevant at these sizes.
public struct Ordered<Value> {

  public private(set) var pairs: [(String, Value)] = []

  public init() {}

  public init(_ entries: [(String, Value)]) {
    for (key, value) in entries {
      self[key] = value
    }
  }

  /// Read or write one entry. Assigning an existing key keeps its position.
  public subscript(key: String) -> Value? {
    get {
      for (entrykey, value) in pairs where entrykey == key {
        return value
      }
      return nil
    }
    set {
      let at = pairs.firstIndex { $0.0 == key }
      if let newvalue = newValue {
        if let at = at {
          pairs[at] = (key, newvalue)
        } else {
          pairs.append((key, newvalue))
        }
      } else if let at = at {
        pairs.remove(at: at)
      }
    }
  }

  public var keys: [String] { return pairs.map { $0.0 } }

  public var values: [Value] { return pairs.map { $0.1 } }

  public var count: Int { return pairs.count }

  public var isEmpty: Bool { return pairs.isEmpty }
}

/// A JSON value.
public indirect enum Json {

  case null
  case bool(Bool)
  case num(Double)
  case str(String)
  case list([Json])
  case map(Ordered<Json>)

  public var asstr: String? {
    if case .str(let value) = self { return value }
    return nil
  }

  public var asnum: Double? {
    if case .num(let value) = self { return value }
    return nil
  }

  public var asbool: Bool? {
    if case .bool(let value) = self { return value }
    return nil
  }

  public var aslist: [Json]? {
    if case .list(let value) = self { return value }
    return nil
  }

  public var asmap: Ordered<Json>? {
    if case .map(let value) = self { return value }
    return nil
  }

  /// Walk nested objects; nil the moment a step is not there.
  public func dig(_ keys: String...) -> Json? {
    return digall(keys)
  }

  public func digall(_ keys: [String]) -> Json? {
    var at: Json? = self

    for key in keys {
      guard let here = at, case .map(let entries) = here else { return nil }
      at = entries[key]
    }

    return at
  }

  /// This value as the text a caller would print, or nil when there is no
  /// value at all. A JSON null is "no value": every provider here treats it
  /// as a miss rather than as the string "null".
  public var text: String? {
    switch self {
    case .null: return nil
    case .str(let value): return value
    case .num(let value): return numstr(value)
    case .bool(let value): return value ? "true" : "false"
    default: return Json.stringify(self)
    }
  }

  // ---- constructors ----

  public static func obj(_ entries: [(String, Json)]) -> Json {
    return .map(Ordered(entries))
  }

  // ---- writing ----

  /// Render a value as compact JSON.
  public static func stringify(_ value: Json) -> String {
    var out = ""
    write(value, &out)
    return out
  }

  /// Render a string as a JSON string literal, quotes included.
  public static func quote(_ text: String) -> String {
    var out = "\""

    for ch in text.unicodeScalars {
      switch ch {
      case "\"": out += "\\\""
      case "\\": out += "\\\\"
      case "\n": out += "\\n"
      case "\r": out += "\\r"
      case "\t": out += "\\t"
      default:
        if 0x20 > ch.value {
          out += String(format: "\\u%04x", ch.value)
        } else {
          out.unicodeScalars.append(ch)
        }
      }
    }

    return out + "\""
  }

  private static func write(_ value: Json, _ out: inout String) {
    switch value {
    case .null:
      out += "null"
    case .bool(let entry):
      out += entry ? "true" : "false"
    case .num(let entry):
      out += numstr(entry)
    case .str(let entry):
      out += quote(entry)
    case .list(let entries):
      out += "["
      for (index, entry) in entries.enumerated() {
        if 0 < index { out += "," }
        write(entry, &out)
      }
      out += "]"
    case .map(let entries):
      out += "{"
      for (index, pair) in entries.pairs.enumerated() {
        if 0 < index { out += "," }
        out += quote(pair.0) + ":"
        write(pair.1, &out)
      }
      out += "}"
    }
  }

  // ---- reading ----

  /// Parse JSON text. nil for anything unreadable - which the caller must
  /// tell apart from a literal `null` body, since only the first means the
  /// store could not answer coherently.
  public static func parse(_ text: String?) -> Json? {
    guard let text = text, !text.isEmpty else { return nil }

    var reader = Reader(Array(text))
    reader.skip()

    guard let value = reader.value(0) else { return nil }

    reader.skip()
    if !reader.done { return nil }

    return value
  }

  /// A recursive-descent reader that answers nil rather than raising: no
  /// error escapes `parse`, because a response body arrives before any
  /// trust check has been made of it.
  private struct Reader {

    // A response body arrives before anything has vouched for it, so the
    // nesting depth is bounded: `[[[[...` must not take the process down
    // with a stack overflow.
    static let MAXDEPTH = 128

    let text: [Character]
    var at: Int = 0

    init(_ text: [Character]) {
      self.text = text
    }

    var done: Bool { return at >= text.count }

    mutating func skip() {
      while !done {
        let ch = text[at]
        if " " == ch || "\t" == ch || "\n" == ch || "\r" == ch {
          at += 1
        } else {
          break
        }
      }
    }

    mutating func value(_ depth: Int) -> Json? {
      if Reader.MAXDEPTH < depth { return nil }
      if done { return nil }

      switch text[at] {
      case "{": return obj(depth)
      case "[": return arr(depth)
      case "\"":
        guard let entry = str() else { return nil }
        return .str(entry)
      case "t":
        return word("true") ? .bool(true) : nil
      case "f":
        return word("false") ? .bool(false) : nil
      case "n":
        return word("null") ? Json.null : nil
      default: return num()
      }
    }

    /// A literal is matched WHOLE: `tru` and `nullx` are not values.
    mutating func word(_ want: String) -> Bool {
      for expect in want {
        guard !done, text[at] == expect else { return false }
        at += 1
      }
      return true
    }

    mutating func obj(_ depth: Int) -> Json? {
      var out = Ordered<Json>()
      at += 1  // {

      skip()
      if !done, "}" == text[at] {
        at += 1
        return .map(out)
      }

      while true {
        skip()
        guard let key = str() else { return nil }
        skip()

        guard !done, ":" == text[at] else { return nil }
        at += 1

        skip()
        guard let entry = value(depth + 1) else { return nil }
        out[key] = entry
        skip()

        if done { return nil }
        if "," == text[at] {
          at += 1
          continue
        }
        if "}" == text[at] {
          at += 1
          return .map(out)
        }
        return nil
      }
    }

    mutating func arr(_ depth: Int) -> Json? {
      var out: [Json] = []
      at += 1  // [

      skip()
      if !done, "]" == text[at] {
        at += 1
        return .list(out)
      }

      while true {
        skip()
        guard let entry = value(depth + 1) else { return nil }
        out.append(entry)
        skip()

        if done { return nil }
        if "," == text[at] {
          at += 1
          continue
        }
        if "]" == text[at] {
          at += 1
          return .list(out)
        }
        return nil
      }
    }

    mutating func str() -> String? {
      guard !done, "\"" == text[at] else { return nil }
      at += 1

      var out = ""

      while !done {
        let ch = text[at]
        at += 1

        if "\"" == ch { return out }

        if "\\" != ch {
          out.append(ch)
          continue
        }

        if done { return nil }
        let escape = text[at]
        at += 1

        switch escape {
        case "\"": out.append("\"")
        case "\\": out.append("\\")
        case "/": out.append("/")
        case "b": out.append("\u{8}")
        case "f": out.append("\u{c}")
        case "n": out.append("\n")
        case "r": out.append("\r")
        case "t": out.append("\t")
        case "u":
          guard at + 4 <= text.count else { return nil }
          let digits = String(text[at..<(at + 4)])
          at += 4
          guard let code = UInt32(digits, radix: 16), let scalar = Unicode.Scalar(code) else {
            return nil
          }
          out.append(Character(scalar))
        default: return nil
        }
      }

      return nil
    }

    mutating func num() -> Json? {
      let start = at

      if !done, "-" == text[at] || "+" == text[at] { at += 1 }

      while !done {
        let ch = text[at]
        if ch.isNumber || "." == ch || "e" == ch || "E" == ch || "-" == ch || "+" == ch {
          at += 1
        } else {
          break
        }
      }

      let span = String(text[start..<at])

      // `1e999` parses to infinity, which JSON has no notation for and
      // which would later blow up a token-expiry computation.
      guard let value = Double(span), value.isFinite else { return nil }

      return .num(value)
    }
  }
}

/// The same reads on an optional value, so a provider can walk a response
/// body - which is `Json?`, because a store may not have answered with JSON
/// at all - without unwrapping at every step.
extension Optional where Wrapped == Json {

  public func dig(_ keys: String...) -> Json? {
    return self?.digall(keys)
  }

  public var text: String? { return self?.text }

  public var asstr: String? { return self?.asstr }

  public var asnum: Double? { return self?.asnum }

  public var asbool: Bool? { return self?.asbool }

  public var aslist: [Json]? { return self?.aslist }

  public var asmap: Ordered<Json>? { return self?.asmap }
}

/// Render a number the way every other port does: a whole number has no
/// fractional tail, so a JSON `1` read back and printed stays `1`.
public func numstr(_ value: Double) -> String {
  if value.isNaN || value.isInfinite { return "null" }

  if value == value.rounded(.towardZero) && 9_007_199_254_740_992.0 > abs(value) {
    return String(Int64(value))
  }

  return String(value)
}
